### PR TITLE
typings attempt

### PR DESCRIPTION
**Pull request by Dor256**
- Adds typings to `Kompot`.
- Adds support for `.ts` test files.

--------

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+type TestComponent = {
+  withMocks(mocks: any[]): TestComponent;
```
Comment (Niryo) [inline index.d.ts]: mocks and triggers are always functions

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+};
+
+export = kompot;
+export as namespace kompot;
+
+declare namespace kompot {
+  function spy(id: string, getReturnValue?: (args: any[]) => any, stringifyArgs?: (args: any[]) => string[]): any
+  function kompotRequire(pathToComponent: string): { [key: string]: TestComponent };
+  function init(): void;
+  function expect<T>(value: T): jest.Matchers<T> & Detox.Expect<T>
```
Comment (Niryo) [inline index.d.ts]: we also have expect spyies

**Review by Niryo — CHANGES_REQUESTED** [review]

File: index.d.ts [hunk 3]
```diff
@@ -7,12 +7,26 @@ type TestComponent = {
...
+
+export = kompot; 
 export as namespace kompot;
 
 declare namespace kompot {
-  function spy(id: string, getReturnValue?: (args: any[]) => any, stringifyArgs?: (args: any[]) => string[]): any
+  function spy(id: string, getReturnValue?: (args: any[]) => any, stringifyArgs?: (args: any[]) => string[]): () => any;
+  function spyOn(obj: object, methodName: string, spyId: string, stringifyArgs?: (args: any[]) => string[]): void;
   function kompotRequire(pathToComponent: string): { [key: string]: TestComponent };
   function init(): void;
```
Comment (Niryo) [inline index.d.ts]: there is a difference between the global kompot and the exported functions from the kompot lib. for example, the init function can only be imported and it's not on the global, and the spyOn function is only on the global.

File: index.d.ts [hunk 4]
```diff
@@ -7,12 +7,26 @@ type TestComponent = {
...
-  function expect<T>(value: T): jest.Matchers<T> & Detox.Expect<T> & {notToHaveBeenCalled(): Promise<void>}
+  function expect<T>(value: T): jest.Matchers<T> & Detox.Expect<T> & {notToHaveBeenCalled(): Promise<void>};
+  function mockFetchUrl(url: string, cb: (url?: string) => Response): void;
```
Comment (Niryo) [inline index.d.ts]: what is cb? anyways, the second parameter is the mock

File: index.d.ts [hunk 5]
```diff
@@ -7,12 +7,26 @@ type TestComponent = {
...
+  function useMocks(getMocks: () => {[key: string]: () => void}): void;
```
Comment (Niryo) [inline index.d.ts]:
```suggestion
  function useMocks(getMocks: () => {[key: string]: () => Promise<void>}): void;
```

Comment (Niryo) [reply index.d.ts]: the mocks can be async

File: index.d.ts [hunk 6]
```diff
@@ -7,12 +7,26 @@ type TestComponent = {
...
 export as namespace kompot;
 
 declare namespace kompot {
-  function spy(id: string, getReturnValue?: (args: any[]) => any, stringifyArgs?: (args: any[]) => string[]): any
+  function spy(id: string, getReturnValue?: (args: any[]) => any, stringifyArgs?: (args: any[]) => string[]): () => any;
+  function spyOn(obj: object, methodName: string, spyId: string, stringifyArgs?: (args: any[]) => string[]): void;
   function kompotRequire(pathToComponent: string): { [key: string]: TestComponent };
   function init(): void;
-  function expect<T>(value: T): jest.Matchers<T> & Detox.Expect<T> & {notToHaveBeenCalled(): Promise<void>}
+  function expect<T>(value: T): jest.Matchers<T> & Detox.Expect<T> & {notToHaveBeenCalled(): Promise<void>};
```
Comment (Niryo) [inline index.d.ts]: kompot expect supports also expect spies

**Review by Niryo — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 6 above)
Comment (Dor256) [reply index.d.ts]: exept for `notToHAveBeenCalled` all of the functions are aligned with the `jest.Matchers` type (TypeScript doesn't care about the runtime evaluation) so this should suffice for the typings.

Either that or I simply don't understand what you mean by `expect spies`

File: index.d.ts (hunk 4 above)
Comment (Dor256) [reply index.d.ts]: forgot to change it - cb stand for callback but it really isn't - will change

File: index.d.ts (hunk 3 above)
Comment (Dor256) [reply index.d.ts]: I see, we might need to ditch the whole `namespace` thing then, let me see

File: index.d.ts (hunk 6 above)
Comment (Niryo) [reply index.d.ts]: something like this:
```javascript
      await expect(global.spy('fetch')).toHaveBeenCalledWith('wow')
```

Comment (Niryo) [reply index.d.ts]: but you are right, I forgot that I used jest api...so it will probably work

Comment (Dor256) [reply index.d.ts]: These types should cover this case

**Review by Niryo — APPROVED** [review]

**Pull request merged**